Controller/Giftcard/AddGiftCard.php

```diff
@@ -0,0 +1,225 @@
+use Magento\Framework\Controller\Result\JsonFactory;
+use Magento\Framework\Exception\LocalizedException;
+use CardknoxDevelopment\Cardknox\Helper\Giftcard;
+use Magento\Checkout\Model\Session as CheckOutSession;
+use Magento\Quote\Api\CartRepositoryInterface;
+use CardknoxDevelopment\Cardknox\Helper\Data;
+
+class AddGiftCard extends Action
+{
+    /**
+     *
+     * @var \Magento\Framework\Controller\Result\JsonFactory
+     */
+    protected $resultJsonFactory;
+
+    /**
+     *
+     * @var \CardknoxDevelopment\Cardknox\Helper\Giftcard
+     */
+    protected $giftcardHelper;
+
+    /**
+     * @var CheckOutSession
+     */
+    protected $checkoutSession;
+
+    /**
+     * @var \Magento\Quote\Api\CartRepositoryInterface
+     */
+    protected $quoteRepository;
+
+    /**
+     * @var \CardknoxDevelopment\Cardknox\Helper\Data
+     */
+    protected $helper;
+
+    /**
+     * __construct function
+     *
+     * @param Context $context
+     * @param JsonFactory $resultJsonFactory
+     * @param Giftcard $giftcardHelper
+     * @param CheckOutSession $checkoutSession
+     * @param CartRepositoryInterface $quoteRepository
+     * @param Data $helper
+     */
+    public function __construct(
+        Context $context,
+        JsonFactory $resultJsonFactory,
+        Giftcard $giftcardHelper,
+        CheckOutSession $checkoutSession,
+        CartRepositoryInterface $quoteRepository,
+        Data $helper
+    ) {
+        $this->resultJsonFactory = $resultJsonFactory;
+        $this->giftcardHelper = $giftcardHelper;
+        $this->checkoutSession = $checkoutSession;
+        $this->quoteRepository = $quoteRepository;
+        $this->helper = $helper;
+        parent::__construct($context);
+    }
+
+    /**
+     * Add GIFT Card
+     *
+     * @return mixed
+     */
+    public function execute()
+    {
+        if (!$this->helper->isCardknoxGiftcardEnabled()) {
+            return $this->jsonResponse(false, __('Please enable Cardknox GiftCard.'));
+        }
+
+        $giftCardCode = $this->getRequest()->getParam('giftcard_code');
+        $isCartPage = $this->getRequest()->getParam('is_cart_page');
+        $selectedShippingMethod = $this->getRequest()->getParam('selected_method');
+
+        if (!$giftCardCode) {
+            return $this->jsonResponse(false, __('Gift Card code is required.'));
+        }
+
+        try {
+            $quote = $this->checkoutSession->getQuote();
+
+            // Validate quote
+            if (!$quote || !is_object($quote)) {
+                return $this->jsonResponse(false, __('We couldn\'t retrieve your cart details. Please try again.'));
+            }
+
+            if ($isCartPage && !$quote->isVirtual()) {
+                $this->updateShippingMethod($quote, $selectedShippingMethod);
+            }
+
+            if ($this->isCartTotalZero($quote)) {
+                return $this->jsonResponse(false, __('Your cart total is zero, so there\'s no need to apply a Gift Card code.'));
```

Comment:
'Your cart total is zero, Gift Card not applicable.'

Reply:
@ybroyde , Updated error message.